akka-stream/src/main/scala/akka/stream/ContextMapStrategy.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream
+
+object ContextMapStrategy {
+
+  final case class Iterate[In, Ctx, Out](
+      iterateFn: (In, Ctx, Out, Int, Boolean) => Ctx,
+      emptyFn: Option[(In, Ctx) => (Out, Ctx)] = None)
+
+  /**
+   * Passthrough the same context [[Ctx] of the input element [[In]] to all output elements [[Out]].
+   */
+  def same[In, Ctx, Out](): Iterate[In, Ctx, Out] = Iterate(
+    iterateFn = (_: In, inCtx: Ctx, _: Out, _: Int, _: Boolean) => inCtx
+  )
+
+  /**
+   * Transform the context of the first element given the input element [[In]] and context [[Ctx]].
+   */
+  def first[In, Ctx, Out](firstFn: (In, Ctx, Out) => Ctx): Iterate[In, Ctx, Out] = Iterate(
+    iterateFn = (in: In, inCtx: Ctx, out: Out, index: Int, _: Boolean) =>
+      if (index == 0) firstFn(in, inCtx, out)
+      else inCtx
+  )
+
+  /**
+   * Transform the context of the last element given the input element [[In]], context [[Ctx]], and the index
+   * of the element.
+   */
+  def last[In, Ctx, Out](fn: (In, Ctx, Out, Int) => Ctx): Iterate[In, Ctx, Out] = Iterate(
+    iterateFn = (in: In, inCtx: Ctx, out: Out, index: Int, hasNext: Boolean) =>
+      if (!hasNext) fn(in, inCtx, out, index)
+      else inCtx
+  )
+
+  /**
+   * Iterate over each element and transform the context given the element [[In]], context [[InCtx]],
+   * output element [[Out]], index, and if the there is a next output element [[Out]]. This strategy can be used to
+   * satisfy all 1:many use cases.
+   */
+  def iterate[In, Ctx, Out](fn: (In, Ctx, Out, Int, Boolean) => Ctx): Iterate[In, Ctx, Out] = Iterate(fn)
```

Comment:
Is the index really needed? Otherwise, it could be `isFirst: Boolean` and `isLast: Boolean` if that's what the index would be used for usually.